update stripe sdk and webhooks to match

vincanger opened the pull request:
## Description

fixes #229 #198 #293 #412 

## Contributor Checklist

> Make sure to do the following steps if they are applicable to your PR:

- [x] **Update e2e tests**: If you changed the [/template/app](/template/app), then make sure to do any neccessary updates to [/template/e2e-tests](/template/e2e-tests) also.
- [x] **Update demo app**: If you changed the [/template/app](/template/app), then make sure to do any neccessary updates to [/opensaas-sh/app_diff](/opensaas-sh/app_diff) also. Check [/opensaas-sh/README.md](/opensaas-sh/README.md) for details.
- [x] **Update docs**: If needed, update the [/opensaas-sh/blog/src/content/docs](/opensaas-sh/blog/src/content/docs).

vincanger commented on template/app/src/payment/stripe/webhook.ts:
```diff
@@ -32,8 +32,8 @@ export const stripeWebhook: PaymentsWebhook = async (request, response, context)
         break;
-      case 'payment_intent.succeeded':
-        await handlePaymentIntentSucceeded(data, prismaUserDelegate);
+      case 'customer.subscription.created':
```
subscription created events used to be sent in customer.subscription.updated events, but they stopped that , so to handle new subscriptions we need this event.

sodic commented:
I'll take a look tomorrow.

sodic commented on template/app/src/payment/plans.ts:
```diff
@@ -5,6 +5,7 @@ export enum SubscriptionStatus {
   CancelAtPeriodEnd = 'cancel_at_period_end',
   Active = 'active',
   Deleted = 'deleted',
+  Pending = 'pending',
```
How come we added this?

vincanger replied:
I added it there as kind of a catch in case they introduce async payment methods or trial periods for their product and forget to update the code. See my other comment below.

sodic reviewed (changes requested):
Hmm, I started reviewing, but got a little confused by the context.

The PR description mentions it closes #229 #198 #293 #412. I went through those issues but couldn't figure out how most of the code in this PR relates to them.

@vincanger is there another issue that this PR addresses? Could you comment on the unexpected code and explain how they relate to those issues, and mention the issue they relate to.

As an example, I understand why the stripe version bump is in this PR, but I don't understand what's happening with the Pending status and the stuff in `extractPriceId`.

vincanger commented on template/app/src/payment/stripe/stripeClient.ts:
```diff
@@ -8,5 +8,5 @@ export const stripe = new Stripe(requireNodeEnvVar('STRIPE_API_KEY'), {
   // For more details and alternative setups check
   // https://docs.stripe.com/api/versioning .
-  apiVersion: '2022-11-15',
+  apiVersion: '2025-04-30.basil',
```
fixes issue #229

vincanger commented on template/app/src/payment/stripe/checkoutUtils.ts:
```diff
@@ -54,33 +51,14 @@ export async function createStripeCheckoutSession({
       success_url: `${DOMAIN}/checkout?success=true`,
       cancel_url: `${DOMAIN}/checkout?canceled=true`,
       automatic_tax: { enabled: true },
+      allow_promotion_codes: true,
```
fixes #412

vincanger commented on template/app/src/payment/stripe/webhook.ts:
```diff
@@ -85,67 +85,60 @@ export const stripeMiddlewareConfigFn: MiddlewareConfigFn = (middlewareConfig) =
+  return updateUserStripePaymentDetails(
+    { userStripeId, numOfCreditsPurchased, datePaid: new Date() },
+    prismaUserDelegate
+  );
```
due to updates to the webhook events on Stripe's side, we can now process credits-based payments and subscriptions within the same endpoint. If it's a subscription payment, `numOfCreditsPurchased` will be undefined and thus won't be updated in the DB. We

vincanger commented on template/app/src/payment/stripe/webhook.ts:
```diff
@@ -85,67 +85,60 @@ export const stripeMiddlewareConfigFn: MiddlewareConfigFn = (middlewareConfig) =
+  const subscriptionStatus: SubscriptionStatus =
+    subscription.status === SubscriptionStatus.Active
+      ? SubscriptionStatus.Active
+      : SubscriptionStatus.Pending;
```
When a new subscription is created, it can be a few options on Stripe's side: `active`, `pending`, `trial`, etc. Async payment types, such as SEPA transfers, would trigger the `pending` type and have to be dealt with in a `checkout.session.async_payment_succeeded` event handler (which we don't include here as most people will just use immediate payments with card). In case they do, the status will be updated with `pending` here and then later will be sent as `completed` in the async handler. Similarly, if they choose to allow for trial periods on their products, they would have to implement that logic to deal with `trials` here.

sodic commented on template/app/src/payment/stripe/webhook.ts:
```diff
@@ -85,69 +85,63 @@ export const stripeMiddlewareConfigFn: MiddlewareConfigFn = (middlewareConfig) =
 ) {
-  const userStripeId = session.customer;
-  const lineItems = await getSubscriptionLineItemsBySessionId(session.id);
+  if (isSuccessfulOneTimePayment(session)) await saveSuccessfulOneTimePayment(session, prismaUserDelegate);
+}
```
This is great, two comments:
- It seems our convention is always putting curly braces around statements in if blocks and put each statement in its own line, so let's do that here. Btw, this is a the better choice according to most anyway (plenty of reasons ranging from readability to maintainability and bugs, I won't list them all here but they're easily discoverable)
- `isSuccessfulOneTimePayment` can be a local variable. It's probably not general enough to be a function.

sodic commented on template/app/src/payment/stripe/webhook.ts:
```diff
@@ -85,69 +85,63 @@ export const stripeMiddlewareConfigFn: MiddlewareConfigFn = (middlewareConfig) =
-  paymentIntent: PaymentIntentSucceededData,
+// This is called when a subscription is successfully created
+// But we wait to update the user's subscription status until the invoice is paid in the handleInvoicePaid function.
+export async function handleCustomerSubscriptionCreated(
```
Can we do the same thing we did in `handleCheckoutSessionCompleted` with these two functions (`handleCustomerSubscriptionCreated` and `handleInvoicePaid`)?

I mean extracting nameable code into something with a name.

sodic reviewed (commented):
Great stuff! I left two suggestions.

vincanger commented:
Alright @sodic I cleaned up and renamed things as you asked, as well as removing `exports` from the handlers since they aren't used anywhere outside the webhook file, and added an `async` declaration for clarity.

sodic reviewed (commented):
Did another pass.

It's taking a little long and I'm noticing new stuff on every iteration because I'm unfamiliar with the context, so thanks for the patience.

Are you up for another call to clear some things up for me?
We can chat tomorrow, To prepare: I'll basically ask why each part of each event handler does stuff the way it does because I can't figure out if the API is weird or if we're not using it right.

Maybe you can point me to the relevant docs and I can take a look? I'm hoping they have a flowchart :)
